index.rst

```diff
@@ -1,57 +1,451 @@
 
-.. Please do not modify tocdepth; will be fixed when a new Sphinx theme is shipped.
+The net result is that a large fraction of our classes must:
 
-.. sectnum::
+- live in ``afw`` itself, leading to churn and bloat at the bottom of our dependency tree, and hence frequent, slow, rebuilds;
 
-.. TODO: Delete the note below before merging new content to the master branch.
+- be implemented in C++, a language much more complex and much less familiar to most of our developers than Python.
 
-.. note::
+Furthermore, the circular dependency between the ``lsst.afw.image`` and ``lsst.afw.table`` subpackages makes the build and particular the Python import order for ``afw`` fragile, leading to hard-to-debug and even harder-to-fix bugs whenever a new :class:`~lsst.afw.image.Exposure`-component class is added.
 
-   **This technote is not yet published.**
+A related problem is that ``lsst.afw.table.io`` was never intended to be a long-term or general persistence solution.\ [#general_persistence]_
+It lacks any kind of built-in support for versioning, produces opaque and sometimes inefficient files, involves a lot of boilerplate to use, and is built on top of a bespoke C++ library (``lsst.afw.table`` itself) that presents a maintenance challenge and has I/O performance problems.
+None of these limitations represents a particularly acute problem (we have managed to work around all of them for a while now), but as a persistence framework more or less specifies the on-disk format, significant (i.e., backwards incompatible) changes will become progressively harder to make as we approach operations.
+
+.. _Community post: https://community.lsst.org/t/how-the-exposure-class-and-afw-io-wrecked-the-codebase/3384
+
+.. [#general_persistence] ``lsst.afw.table.io`` was written to enable writing :class:`~lsst.meas.algorithms.CoaddPsf` objects into FITS files on an HSC data release deadline, with the expectation that it would later be replaced by the Data Access Team.
+   That "Data Access" has since been reinterpreted to mean only *science user* data access (i.e., not pipeline data access or data access middleware), and the hence the responsibility for persistence frameworks has fallen through the cracks.
+
+
+.. _overview:
+
+Overview
+========
+
+We've taken the following actions to mitigate the design problems imposed by :class:`~lsst.afw.image.Exposure` and to prepare for future work:
+
+- Add a new heterogeneous-type mapping interface and implementation in C++, :class:`~lsst.afw.typehandling.GenericMap`.
+  This augments rather than replaces our existing :class:`~lsst.daf.base.PropertySet` and :class:`~lsst.daf.base.PropertyList` to avoid their behavioral idiosyncrasies (especially with respect to vector-valued entries) and to provide natural interfaces in both C++ and Python (e.g., the :class:`collections.abc.Mapping` API).
```

Comment:
I would read “augment” here as saying that `GenericMap' builds atop the daf_base classes (by subclassing, code re-use, or similar). I don't think that's the case (is it?). Rather the stack has been augmented by adding a new, independent implementation of some of the same same capabilities.